File: Algorithm/C++/1000+/1365.how-many-numbers-are-smaller-than-the-current-number.cpp
/*
 * @lc app=leetcode id=1365 lang=cpp
 *
 * [1365] How Many Numbers Are Smaller Than the Current Number
 */

// @lc code=start
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    std::vector<int> smallerNumbersThanCurrent(std::vector<int>& nums) {
        std::vector<int> original_nums(nums);
        std::sort(nums.begin(), nums.end());
        std::unordered_map<int, int> counter;
        for (int i = 0; i < nums.size(); i++) {
            if (counter.find(nums[i]) == counter.end()) {
                counter[nums[i]] = i;
            }
        }
        std::vector<int> result;
        result.reserve(nums.size());
        for (int num: original_nums) {
            result.push_back(counter[num]);
        }
        return result;
    }
};
// @lc code=end

File: Algorithm/Python/1000+/1365.how-many-numbers-are-smaller-than-the-current-number.py
#
# @lc app=leetcode id=1365 lang=python
#
# [1365] How Many Numbers Are Smaller Than the Current Number
#

# @lc code=start
class Solution(object):
    def smallerNumbersThanCurrent(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        sorted_nums = sorted(nums)
        counter = {}
        for i in range(len(sorted_nums)):
            if sorted_nums[i] not in counter:
                counter[sorted_nums[i]] = i
        return [counter[num] for num in nums] 
    
# @lc code=end

